plugins-compat-tester-model/src/main/java/org/jenkins/tools/test/model/hook/PluginCompatTesterHooks.java

```diff
@@ -95,63 +111,79 @@ public PluginCompatTesterHooks(List<String> extraPrefixes) {
-        for(String prefix : hookPrefixes) {
-            Reflections reflections = new Reflections(prefix);
-            Set<Class<? extends PluginCompatTesterHook>> subTypes;
-            
-            // Find all steps for a given stage. Long due to casting
-            switch(stage) {
-                case "compilation" : 
-                    Set<Class<? extends PluginCompatTesterHookBeforeCompile>> compSteps = reflections.getSubTypesOf(PluginCompatTesterHookBeforeCompile.class); 
-                    subTypes = compSteps.stream()
-                        .map(this::casting)
-                        .collect(Collectors.toSet());
-                    break;
-                case "execution" : 
-                    Set<Class<? extends PluginCompatTesterHookBeforeExecution>> exeSteps = reflections.getSubTypesOf(PluginCompatTesterHookBeforeExecution.class); 
-                    subTypes = exeSteps.stream()
-                        .map(this::casting)
-                        .collect(Collectors.toSet());
-                    break;
-                case "checkout" : 
-                    Set<Class<? extends PluginCompatTesterHookBeforeCheckout>> checkSteps = reflections.getSubTypesOf(PluginCompatTesterHookBeforeCheckout.class); 
-                    subTypes = checkSteps.stream()
-                        .map(this::casting)
-                        .collect(Collectors.toSet());
-                    break;
-                default: // Not valid; nothing will get executed
-                    return new HashMap<>();
-            }
-            
-            for(Class<?> c : subTypes) {
-                // Ignore abstract hooks
-                if (!Modifier.isAbstract(c.getModifiers())) {
-                    try {
-                        System.out.println("Hook: " + c.getName());
-                        Constructor<?> constructor = c.getConstructor();
-                        PluginCompatTesterHook hook = (PluginCompatTesterHook) constructor.newInstance();
-
-                        List<String> plugins = hook.transformedPlugins();
-                        for (String plugin : plugins) {
-                            Queue<PluginCompatTesterHook> allForType = sortedHooks.get(plugin);
-                            if (allForType == null) {
-                                allForType = new LinkedList<>();
-                            }
-                            allForType.add(hook);
-                            sortedHooks.put(plugin, allForType);
+        Reflections reflections = new Reflections(hookPrefixes.toArray(new String[0]), classLoaders.toArray(new ClassLoader[classLoaders.size()]));
+        Set<Class<? extends PluginCompatTesterHook>> subTypes;
+
+        // Find all steps for a given stage. Long due to casting
+        switch(stage) {
+        case "compilation" : 
+            Set<Class<? extends PluginCompatTesterHookBeforeCompile>> compSteps = reflections.getSubTypesOf(PluginCompatTesterHookBeforeCompile.class); 
```

Comment:
Just found this PR. Is it too late to request that this be improved to use either `ServiceLoader` (with `metainf-services-annotation`) or https://github.com/jglick/sezpoz? Scanning bytecode for classes that happen to implement some interface is pretty fragile (as well as more work than the clean solutions).

Reply:
I suggested the former

Reply:
I understand that PCT was already using bytecode scanning, but this PR switches it from an unfortunate implementation detail to a public API.